FAI-678 DfESignIn NuGet

alireid opened the pull request.

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Helpers/JsonWebAlgorithmTests.cs:
```diff
@@ -0,0 +1,39 @@
+        [SetUp]
+        public void SetUp()
+        {
+            this.mockRepository = new MockRepository(MockBehavior.Strict);
```
no `this`

alireid replied:
Thanks Dan, removed all those

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Client/DfESignInClientFactory.cs:
```diff
@@ -0,0 +1,43 @@
+
+        public DfESignInClient CreateDfESignInClient(string userId = "", string organisationId = "")
+        {
+            var dsiConfiguration = _config.GetSection("DfEOidcConfiguration").Get<DfEOidcConfiguration>();
```
i would have thought we could have just resolved the config in the DI rather than doing this?

alireid replied:
Good spot Dan, not sure why that was left like that as config all ready fired up in DI. Fixed up, thanks.

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Client/DfESignInClientFactory.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            var dsiConfiguration = _config.GetSection("DfEOidcConfiguration").Get<DfEOidcConfiguration>();
+
+            var dfeSignInClient = new DfESignInClient(new HttpClient())
```
dont really want to be new'ing up the HttpClient - this should be injected

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Client/DfESignInClientFactory.cs:
```diff
@@ -0,0 +1,43 @@
+                .Issuer(dsiConfiguration.ClientId)
+                .CreateToken();
+
+            dfeSignInClient.HttpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {token}");
```
using the DefaultRequestHeader is not thread safe and you risk having multiple headers added. Should also be doing:
```
.Authorization = new AuthenticationHeaderValue("Bearer", token);
```

alireid replied:
Nice, fixed that up

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Client/DfESignInClientFactory.cs:
```diff
@@ -0,0 +1,43 @@
+
+            var tokenData = new TokenData();
+            tokenData.Header.Add("typ", "JWT");
+            var token = new TokenBuilder(new TokenDataSerializer(), tokenData, new TokenEncoder(), new JsonWebAlgorithm())
```
This seems like a lot of new that is wrong all this should be injected?

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Helpers/TokenBuilder.cs:
```diff
@@ -0,0 +1,58 @@
+        public string Audience { get; set; }
+        public byte[] SecretKey { get; set; }
+        public string Algorithm { get; set; }
+        public ITokenData TokenData { get; set; }
```
Shouldn't this be private?

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Helpers/TokenBuilder.cs:
```diff
@@ -0,0 +1,58 @@
+
+        public string CreateToken()
+        {
+            if (string.IsNullOrWhiteSpace(Algorithm)) throw new Exception("Algorithm");
+            if (string.IsNullOrWhiteSpace(Issuer)) throw new Exception("Issuer");
+            if (string.IsNullOrWhiteSpace(Audience)) throw new Exception("Audience");
+            if (SecretKey == null || SecretKey.Length < 1) throw new Exception("SecretKey");
+
+            byte[] headerBytes = Encoding.UTF8.GetBytes(_tokenDataSerializer.Serialize(TokenData.Header));
+            byte[] payloadBytes = Encoding.UTF8.GetBytes(_tokenDataSerializer.Serialize(TokenData.Payload));
+            byte[] bytesToSign = Encoding.UTF8.GetBytes($"{_tokenEncoder.Base64Encode(headerBytes)}.{_tokenEncoder.Base64Encode(payloadBytes)}");
+            byte[] signedBytes = SignToken(SecretKey, bytesToSign);
+
+            string token = $"{_tokenEncoder.Base64Encode(headerBytes)}.{_tokenEncoder.Base64Encode(payloadBytes)}.{_tokenEncoder.Base64Encode(signedBytes)}";
+
+            return token;
```
I feel like this is missing test coverage

alireid replied:
Agreed there Dan, thanks, didn't spent too much time with the unit tests, adding a load in now

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Api/Helpers/TokenBuilderExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+{
+    public static class TokenBuilderExtensions
+    {
+        public static TokenBuilder ForAudience(this TokenBuilder tokenBuilder, string audience)
```
I think this should be TokenData extensions rather than TokenBuilder and expose the insides

balaji-j replied:
Hi Dan, As per PR recommendations I have modified the code base and these are the modifications below 

1) Renamed file TokenBuilderExtension to TokenDataExtensions
2) Amended Exception messages to appropriate ones
3) Included Unit Test cases to cover the logic under TokenDataExtensions

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/AppStart/AddServiceRegistrationExtension.cs:
```diff
@@ -0,0 +1,32 @@
+            }
+
+            services.AddOptions();
+            services.AddTransient(typeof(ICustomClaims), customClaims);
```
Not sure we need CustomClaims for this one? it looks like its only used in the stub

alireid replied:
Discovered this one when I included it another project, good shout, thanks, removed!

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Configuration/DfEOidcConfiguration.cs:
```diff
@@ -0,0 +1,22 @@
+        public string APIServiceSecret { get; set; }
+    }
+
+    public interface IDfEOidcConfiguration
```
Do we need this?

alireid replied:
Agreed, surplus! Removed, thanks

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/AppStart/ProviderStubAuthHandler.cs:
```diff
@@ -0,0 +1,51 @@
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            var claims = new List<Claim>
+            {
+                new Claim(ClaimTypes.Email, _configuration["NoAuthEmail"]),
+                new Claim(ClaimTypes.NameIdentifier, Guid.Empty.ToString()),
+                new Claim("sub", Guid.Empty.ToString())
+            };
+
+            if (_customClaims != null)
+            {
+                var additionalClaims = await _customClaims.GetClaims(null);
+                claims.AddRange(additionalClaims);
+            }
+
+            var identity = new ClaimsIdentity(claims, "Provider-stub");
+            var principal = new ClaimsPrincipal(identity);
+            var ticket = new AuthenticationTicket(principal, "Provider-stub");
+
+            var result = AuthenticateResult.Success(ticket);
+
+            return result;
```
this probably needs to be more geared towards provider auth, adding the ukprn and necessary role

alireid replied:
Thanks Dan, didn't know what claims were needed until first implementation, added in now.

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/AppStart/ConfigureDfESignInAuthenticationExtension.cs:
```diff
@@ -0,0 +1,146 @@
+            
+            var ukPrn = userOrganisation.UkPrn ?? 10000531;
+
+            await DfEPublicApi(ctx, userId, userOrganisation.Id.ToString(), config);
```
i feel like this should be injected rather than new'd up inside the method

alireid replied:
Thanks Dan, nice - got that in

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Client/DfESignInClientFactoryTests.cs:
```diff
@@ -0,0 +1,45 @@
+    [TestFixture]
+    public class DfESignInClientFactoryTests
+    {
+        private MockRepository mockRepository;
```
Good spot, thanks Dan, fixed those up

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Client/DfESignInClientFactoryTests.cs:
```diff
@@ -0,0 +1,45 @@
+        [SetUp]
+        public void SetUp()
+        {
+            this.mockRepository = new MockRepository(MockBehavior.Strict);
```
Good spot, thanks Dan, fixed those up

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Client/DfESignInClientFactoryTests.cs:
```diff
@@ -0,0 +1,45 @@
+        public void CreateDfESignInClient_StateUnderTest_ExpectedBehavior()
+        {
+            // Arrange
+            var factory = this.CreateFactory();
```
Agreed, removed!

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Helpers/JsonWebAlgorithmTests.cs:
```diff
@@ -0,0 +1,39 @@
+    [TestFixture]
+    public class JsonWebAlgorithmTests
+    {
+        private MockRepository mockRepository;
```
Fixed up those, thanks

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Configuration/DfEOidcConfigurationTests.cs:
```diff
@@ -0,0 +1,36 @@
+
+            // Assert
+            this.mockRepository.VerifyAll();
+            Assert.That(dfEOidcConfiguration, Is.Not.Null);
```
Ha yes, agreed, nuked!

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Helpers/JsonWebAlgorithmTests.cs:
```diff
@@ -0,0 +1,39 @@
+            this.mockRepository = new MockRepository(MockBehavior.Strict);
+        }
+
+        private JsonWebAlgorithm CreateJsonWebAlgorithm()
```
Thanks Dan, agreed, cleaned up that up

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Helpers/JsonWebAlgorithmTests.cs:
```diff
@@ -0,0 +1,39 @@
+
+            // Assert
+            this.mockRepository.VerifyAll();
+            Assert.That(result, Is.Not.Null);
```
Correct, thanks

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth.UnitTests/Api/Client/DfESignInClientFactoryTests.cs:
```diff
@@ -0,0 +1,45 @@
+            var mockIConfigurationSection = new Mock<IConfigurationSection>();
+            mockIConfigurationSection.Setup(x => x.GetSection("DfEOidcConfiguration")).Returns(mockIConfigurationSection.Object);
+
+            // Assert
```
Thanks Dan, agreed, working on unit test now - added a load in

dashton82 commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/AppStart/AddServiceRegistrationExtension.cs:
```diff
@@ -0,0 +1,40 @@
+    internal static class AddServiceRegistrationExtension
+    {
+        internal static void AddServiceRegistration(this IServiceCollection services, IConfiguration configuration,
+            Type customClaims)
```
This can be removed as it isn't used anymore

scottywakefield reviewed (commented):
Hi @alireid, could you please use cdn sources for the libs in the sample site plz, will remove about 39,000 lines of code from the PR and the repo.

alireid commented:
> Hi @alireid, could you please use cdn sources for the libs in the sample site plz, will remove about 39,000 lines of code from the PR and the repo.

Hi @scottywakefield - great shout, removed and added cdns.

alireid commented on SFA.DAS.DfESignIn.Auth/SFA.DAS.DfESignIn.Auth/Interfaces/IDfESignInClient.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace SFA.DAS.DfESignIn.Auth.Interfaces
+{
+    public interface IDfESignInClient
```
Agreed, surplus, removed
